### PR TITLE
[FEAT]: Waitlist API - BACKEND

### PR DESCRIPTION
**Pull request by gurumarv**
This pull request adds the following:
- `EmailProducerService`: Service for sending email messages.
- `EmailConsumerService`: Service for receiving and processing email messages.
- `RabbitMQConfig`: Configuration for RabbitMQ.
- `EmailMessage`: Entity class for email messages.

### **Description**
Create a backend API endpoint to receive and store user information submitted through the waitlist signup form on the website.

### **Acceptance Criteria**

- The endpoint should be accessible at /api/v1/waitlist.

- The endpoint should accept POST requests to receive user information.

- Given a POST request with user data, the system should validate and sanitize the incoming data.

- Validation rules should include:

  -   Email format validation.
  -   Required fields: email, name.
  -   The name should be a non-empty string.

-  Given valid user data, the system should securely store the information in the database.

Waitlist API

- Upon successful signup, the system should send a confirmation email to the user.

- Return a 429 Too Many Requests status with an appropriate error message if the rate limit is exceeded

- The system should handle errors gracefully and return appropriate status codes.

Request Body

```
{
  "email": "johndoe@gmail.com",
  "fullName": "John Doe"
}
```

Successful Response (status code: 201)

```
{
  "message": "You are all signed up!",
}
```

Error Response

```
{
  "message": "Invalid email format",
  "status_code": 400,
  "error": "Bad Request"
}
```

### **Purpose**
To allow potential users to join a waitlist for our product/service by submitting their information through a form on the website.

### **Requirements**
 

- [x] Develop a RESTful API endpoint to handle POST requests
- [x]  Validate and sanitize incoming user data
- [x]  Store user information securely in the database
- [x]  Implement rate limiting to prevent abuse
- [x]  Send a confirmation email to the user upon successful signup
- [x]  Handle errors gracefully and return appropriate status codes

### **Expected Outcome**
A secure and efficient API endpoint that can receive user information, store it in the database and confirm the user's addition to the waitlist.

### **Testing**

- Test that the endpoint returns a 201 status code with a success message on valid submissions.
- Test that invalid data triggers a 400 Bad Request response with an appropriate error message.
- Test that a confirmation email is sent upon successful signup.
- Test that the rate-limiting works correctly and returns the appropriate status code when exceeded.

### CODE REVIEW

File: src/main/java/hng_java_boilerplate/config/RabbitMQConfig.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import org.springframework.amqp.core.Queue;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class RabbitMQConfig {
+    @Bean
+    public Queue emailQueue() {
+        return new Queue("emailQueue", true);
```
Comment (billmal071) [inline src/main/java/hng_java_boilerplate/config/RabbitMQConfig.java:11]: use application.properties file for the queue name

File: src/main/java/hng_java_boilerplate/waitlist/controller/WaitlistController.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+@RestController
+@RequestMapping("api/v1/waitlist")
+public class WaitlistController {
+    @Autowired
+    private WaitlistService waitlistService;
+    @Autowired
+    private EmailProducerService emailProducerService;
```
Comment (billmal071) [inline src/main/java/hng_java_boilerplate/waitlist/controller/WaitlistController.java:25]: Make this consistent with your other fields. Prefer constructor injection.

File: src/main/java/hng_java_boilerplate/waitlist/entity/Waitlist.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+@Entity
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+public class Waitlist {
+    @Id
+    @GeneratedValue(strategy = GenerationType.UUID)
+    private UUID id;
+    @NotBlank(message = "Email is required")
```
Comment (billmal071) [inline src/main/java/hng_java_boilerplate/waitlist/entity/Waitlist.java:26]: fullName

**Review by billmal071 — CHANGES_REQUESTED** [review]

File: src/main/java/hng_java_boilerplate/config/RabbitMQConfig.java (hunk 1 above)
Comment (gurumarv) [reply src/main/java/hng_java_boilerplate/config/RabbitMQConfig.java]: Done. Thanks

File: src/main/java/hng_java_boilerplate/waitlist/controller/WaitlistController.java (hunk 2 above)
Comment (gurumarv) [reply src/main/java/hng_java_boilerplate/waitlist/controller/WaitlistController.java]: Done. Thanks

File: src/main/java/hng_java_boilerplate/email/EmailServices/EmailConsumerService.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+@Service
+public class EmailConsumerService {
+    private final JavaMailSender emailSender;
+
+    @Autowired
+    public EmailConsumerService(JavaMailSender emailSender) {
+        this.emailSender = emailSender;
+    }
+
+    @RabbitListener(queues = "emailQueue")
```
Comment (billmal071) [inline src/main/java/hng_java_boilerplate/email/EmailServices/EmailConsumerService.java:19]: Use the variable in the properties file.

File: src/main/java/hng_java_boilerplate/email/EmailServices/EmailProducerService.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    private final RabbitTemplate rabbitTemplate;
+
+    @Autowired
+    public EmailProducerService(RabbitTemplate rabbitTemplate) {
+        this.rabbitTemplate = rabbitTemplate;
+    }
+
+    public void sendEmailMessage(String to, String subject, String text) {
+        EmailMessage emailMessage = new EmailMessage(to, subject, text);
+        rabbitTemplate.convertAndSend("emailQueue", emailMessage);
```
Comment (billmal071) [inline src/main/java/hng_java_boilerplate/email/EmailServices/EmailProducerService.java:19]: Use the properties file for the queueName.

File: src/main/java/hng_java_boilerplate/waitlist/service/EmailService.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.mail.SimpleMailMessage;
+import org.springframework.mail.javamail.JavaMailSender;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class EmailService {
+
+    @Autowired
+    private JavaMailSender emailSender;
```
Comment (billmal071) [inline src/main/java/hng_java_boilerplate/waitlist/service/EmailService.java:12]: Use constructor injection

File: src/main/java/hng_java_boilerplate/waitlist/service/WaitlistService.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+import hng_java_boilerplate.waitlist.entity.Waitlist;
+import hng_java_boilerplate.waitlist.repository.WaitlistRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class WaitlistService {
+    @Autowired
+    private WaitlistRepository waitlistRepository;
```
Comment (billmal071) [inline src/main/java/hng_java_boilerplate/waitlist/service/WaitlistService.java:11]: Use constructor injection

**Review by billmal071 — CHANGES_REQUESTED** [review]

File: src/main/java/hng_java_boilerplate/email/EmailServices/EmailConsumerService.java (hunk 4 above)
Comment (gurumarv) [reply src/main/java/hng_java_boilerplate/email/EmailServices/EmailConsumerService.java]: Done

File: src/main/java/hng_java_boilerplate/email/EmailServices/EmailProducerService.java (hunk 5 above)
Comment (gurumarv) [reply src/main/java/hng_java_boilerplate/email/EmailServices/EmailProducerService.java]: Done

File: src/main/java/hng_java_boilerplate/waitlist/service/EmailService.java (hunk 6 above)
Comment (gurumarv) [reply src/main/java/hng_java_boilerplate/waitlist/service/EmailService.java]: Done

File: src/main/java/hng_java_boilerplate/waitlist/service/WaitlistService.java (hunk 7 above)
Comment (gurumarv) [reply src/main/java/hng_java_boilerplate/waitlist/service/WaitlistService.java]: Done

**Pull request closed without merging**